File: com.trollworks.gcs/src/com/trollworks/gcs/skill/SkillDefault.java
/*
 * Copyright ©1998-2021 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.skill;

import com.trollworks.gcs.attribute.AttributeDef;
import com.trollworks.gcs.datafile.DataFile;
import com.trollworks.gcs.ui.widget.outline.ListRow;
import com.trollworks.gcs.utility.I18n;
import com.trollworks.gcs.utility.json.JsonMap;
import com.trollworks.gcs.utility.json.JsonWriter;

import java.io.IOException;
import java.util.Map;
import java.util.Set;

/** Describes a skill default. */
public class SkillDefault {
    public static final String KEY_TYPE           = "type";
    public static final String KEY_NAME           = "name";
    public static final String KEY_SPECIALIZATION = "specialization";
    public static final String KEY_MODIFIER       = "modifier";
    public static final String KEY_LEVEL          = "level";
    public static final String KEY_ADJ_LEVEL      = "adjusted_level";
    public static final String KEY_POINTS         = "points";

    private String mType;
    private String mName;
    private String mSpecialization;
    private int    mModifier;
    private int    mLevel;
    private int    mAdjLevel;
    private int    mPoints;

    /**
     * Creates a new skill default.
     *
     * @param type           The type of default.
     * @param name           The name of the skill to default from. Pass in {@code null} if type is
     *                       not skill-based.
     * @param specialization The specialization of the skill. Pass in {@code null} if this does not
     *                       default from a skill or the skill doesn't require a specialization.
     * @param modifier       The modifier to use.
     */
    public SkillDefault(String type, String name, String specialization, int modifier) {
        setType(type);
        setName(name);
        setSpecialization(specialization);
        setModifier(modifier);
    }

    /**
     * Creates a clone of the specified skill default.
     *
     * @param other The skill default to clone.
     */
    public SkillDefault(SkillDefault other) {
        mType = other.mType;
        mName = other.mName;
        mSpecialization = other.mSpecialization;
        mModifier = other.mModifier;
    }

    /**
     * Creates a skill default.
     *
     * @param m    The {@link JsonMap} to load data from.
     * @param full {@code true} if all fields should be loaded.
     */
    public SkillDefault(JsonMap m, boolean full) {
        setType(m.getString(DataFile.KEY_TYPE));
        mName = m.getString(KEY_NAME);
        mSpecialization = m.getString(KEY_SPECIALIZATION);
        mModifier = m.getInt(KEY_MODIFIER);
        if (full) {
            mLevel = m.getInt(KEY_LEVEL);
            mAdjLevel = m.getInt(KEY_ADJ_LEVEL);
            mPoints = m.getInt(KEY_POINTS);
        }
    }

    /** @return The current level of this default. Temporary storage only. */
    public int getLevel() {
        return mLevel;
    }

    /**
     * @param level Sets the current level of this default. Temporary storage only.
     */
    public void setLevel(int level) {
        mLevel = level;
    }

    /** @return The current level of this default. Temporary storage only. */
    public int getAdjLevel() {
        return mAdjLevel;
    }

    /**
     * @param level Sets the current level of this default. Temporary storage only.
     */
    public void setAdjLevel(int level) {
        mAdjLevel = level;
    }

    /**
     * @return The current points provided by this default. Temporary storage only.
     */
    public int getPoints() {
        return mPoints;
    }

    /**
     * @param points Sets the current points provided by this default. Temporary storage only.
     */
    public void setPoints(int points) {
        mPoints = points;
    }

    /**
     * Saves the skill default.
     *
     * @param w    The {@link JsonWriter} to use.
     * @param full {@code true} if all fields should be saved.
     */
    public void save(JsonWriter w, boolean full) throws IOException {
        w.startMap();
        w.keyValue(DataFile.KEY_TYPE, mType);
        if (SkillDefaultType.isSkillBased(mType)) {
            w.keyValueNot(KEY_NAME, mName, "");
            w.keyValueNot(KEY_SPECIALIZATION, mSpecialization, "");
        }
        w.keyValueNot(KEY_MODIFIER, mModifier, 0);
        if (full) {
            w.keyValue(KEY_LEVEL, mLevel);
            w.keyValue(KEY_ADJ_LEVEL, mAdjLevel);
            w.keyValueNot(KEY_POINTS, mPoints, 0);
        }
        w.endMap();
    }

    /** @return The type of default. */
    public String getType() {
        return mType;
    }

    /** @param type The new type. */
    public void setType(String type) {
        mType = AttributeDef.sanitizeID(type, false);
    }

    /** @return The full name of the skill to default from. */
    public String getFullName(DataFile dataFile) {
        if (SkillDefaultType.isSkillBased(mType)) {
            StringBuilder builder = new StringBuilder();
            builder.append(mName);
            if (!mSpecialization.isEmpty()) {
                builder.append(" (");
                builder.append(mSpecialization);
                builder.append(')');
            }
            if ("parry".equalsIgnoreCase(mType)) {
                builder.append(I18n.Text(" Parry"));
            } else if ("block".equalsIgnoreCase(mType)) {
                builder.append(I18n.Text(" Block"));
            }
            return builder.toString();
        }
        return Skill.resolveAttributeName(dataFile, mType);
    }

    /**
     * @return The name of the skill to default from. Only valid when {@link #getType()} returns a
     *         {@link SkillDefaultType} which is skill based.
     */
    public String getName() {
        return mName;
    }

    /** @param name The new name. */
    public void setName(String name) {
        mName = name != null ? name : "";
    }

    /**
     * @return The specialization of the skill to default from. Only valid when {@link #getType()}
     *         returns a {@link SkillDefaultType} which is skill based.
     */
    public String getSpecialization() {
        return mSpecialization;
    }

    /** @param specialization The new specialization. */
    public void setSpecialization(String specialization) {
        mSpecialization = specialization != null ? specialization : "";
    }

    /** @return The modifier. */
    public int getModifier() {
        return mModifier;
    }

    /** @param modifier The new modifier. */
    public void setModifier(int modifier) {
        mModifier = modifier;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj instanceof SkillDefault) {
            SkillDefault sd = (SkillDefault) obj;
            return mModifier == sd.mModifier && mType.equals(sd.mType) && mName.equals(sd.mName) && mSpecialization.equals(sd.mSpecialization);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    /** @param set The nameable keys. */
    public void fillWithNameableKeys(Set<String> set) {
        ListRow.extractNameables(set, getName());
        ListRow.extractNameables(set, getSpecialization());
    }

    /** @param map The map of nameable keys to names to apply. */
    public void applyNameableKeys(Map<String, String> map) {
        setName(ListRow.nameNameables(map, getName()));
        setSpecialization(ListRow.nameNameables(map, getSpecialization()));
    }

    public String getModifierAsString() {
        if (mModifier > 0) {
            return " + " + mModifier;
        } else if (mModifier < 0) {
            return " - " + -mModifier;
        }
        return "";
    }
}

File: model/skill_default.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package model

import (
	"strings"

	"github.com/richardwilkes/gcs/v5/model/fxp"
	"github.com/richardwilkes/toolbox/i18n"
)

var skillBasedDefaultTypes = map[string]bool{
	SkillID: true,
	ParryID: true,
	BlockID: true,
}

// SkillDefault holds data for a Skill default.
type SkillDefault struct {
	DefaultType    string  `json:"type"`
	Name           string  `json:"name,omitempty"`
	Specialization string  `json:"specialization,omitempty"`
	Modifier       fxp.Int `json:"modifier,omitempty"`
	Level          fxp.Int `json:"level,omitempty"`
	AdjLevel       fxp.Int `json:"adjusted_level,omitempty"`
	Points         fxp.Int `json:"points,omitempty"`
}

// DefaultTypeIsSkillBased returns true if the SkillDefault type is Skill-based.
func DefaultTypeIsSkillBased(skillDefaultType string) bool {
	return skillBasedDefaultTypes[strings.ToLower(strings.TrimSpace(skillDefaultType))]
}

// CloneWithoutLevelOrPoints creates a copy, but without the level or points set.
func (s *SkillDefault) CloneWithoutLevelOrPoints() *SkillDefault {
	clone := *s
	clone.Level = 0
	clone.AdjLevel = 0
	clone.Points = 0
	return &clone
}

// Equivalent returns true if this can be considered equivalent to other.
func (s *SkillDefault) Equivalent(other *SkillDefault) bool {
	return other != nil && s.DefaultType == other.DefaultType && s.Modifier == other.Modifier && s.Name == other.Name &&
		s.Specialization == other.Specialization
}

// Type returns the type of the SkillDefault.
func (s *SkillDefault) Type() string {
	return s.DefaultType
}

// SetType sets the type of the SkillDefault.
func (s *SkillDefault) SetType(t string) {
	s.DefaultType = SanitizeID(t, true)
}

// FullName returns the full name of the skill to default from.
func (s *SkillDefault) FullName(entity *Entity) string {
	if s.SkillBased() {
		var buffer strings.Builder
		buffer.WriteString(s.Name)
		if s.Specialization != "" {
			buffer.WriteString(" (")
			buffer.WriteString(s.Specialization)
			buffer.WriteByte(')')
		}
		switch {
		case strings.EqualFold(DodgeID, s.DefaultType):
			buffer.WriteString(i18n.Text(" Dodge"))
		case strings.EqualFold(ParryID, s.DefaultType):
			buffer.WriteString(i18n.Text(" Parry"))
		case strings.EqualFold(BlockID, s.DefaultType):
			buffer.WriteString(i18n.Text(" Block"))
		}
		return buffer.String()
	}
	return ResolveAttributeName(entity, s.DefaultType)
}

// FillWithNameableKeys adds any nameable keys found in this SkillDefault to the provided map.
func (s *SkillDefault) FillWithNameableKeys(m map[string]string) {
	Extract(s.Name, m)
	Extract(s.Specialization, m)
}

// ApplyNameableKeys replaces any nameable keys found in this SkillDefault with the corresponding values in the provided
// map.
func (s *SkillDefault) ApplyNameableKeys(m map[string]string) {
	s.Name = Apply(s.Name, m)
	s.Specialization = Apply(s.Specialization, m)
}

// ModifierAsString returns the modifier as a string suitable for appending.
func (s *SkillDefault) ModifierAsString() string {
	if s.Modifier != 0 {
		return s.Modifier.StringWithSign()
	}
	return ""
}

// SkillBased returns true if the Type() is Skill-based.
func (s *SkillDefault) SkillBased() bool {
	return skillBasedDefaultTypes[strings.ToLower(strings.TrimSpace(s.DefaultType))]
}

// SkillLevel returns the base skill level for this SkillDefault.
func (s *SkillDefault) SkillLevel(entity *Entity, requirePoints bool, excludes map[string]bool, ruleOf20 bool) fxp.Int {
	switch s.Type() {
	case ParryID:
		best := s.best(entity, requirePoints, excludes)
		if best != fxp.Min {
			best = best.Div(fxp.Two).Trunc() + fxp.Three + entity.ParryBonus
		}
		return s.finalLevel(best)
	case BlockID:
		best := s.best(entity, requirePoints, excludes)
		if best != fxp.Min {
			best = best.Div(fxp.Two).Trunc() + fxp.Three + entity.BlockBonus
		}
		return s.finalLevel(best)
	case SkillID:
		return s.finalLevel(s.best(entity, requirePoints, excludes))
	default:
		return s.SkillLevelFast(entity, requirePoints, excludes, ruleOf20)
	}
}

func (s *SkillDefault) best(entity *Entity, requirePoints bool, excludes map[string]bool) fxp.Int {
	best := fxp.Min
	for _, sk := range entity.SkillNamed(s.Name, s.Specialization, requirePoints, excludes) {
		if best < sk.LevelData.Level {
			level := sk.CalculateLevel().Level
			if best < level {
				best = level
			}
		}
	}
	return best
}

// SkillLevelFast returns the base skill level for this SkillDefault.
func (s *SkillDefault) SkillLevelFast(entity *Entity, requirePoints bool, excludes map[string]bool, ruleOf20 bool) fxp.Int {
	switch s.Type() {
	case DodgeID:
		level := entity.Dodge(entity.EncumbranceLevel(false))
		if ruleOf20 && level > 20 {
			level = 20
		}
		return s.finalLevel(fxp.From(level))
	case ParryID:
		best := s.bestFast(entity, requirePoints, excludes)
		if best != fxp.Min {
			best = best.Div(fxp.Two).Trunc() + fxp.Three + entity.ParryBonus
		}
		return s.finalLevel(best)
	case BlockID:
		best := s.bestFast(entity, requirePoints, excludes)
		if best != fxp.Min {
			best = best.Div(fxp.Two).Trunc() + fxp.Three + entity.BlockBonus
		}
		return s.finalLevel(best)
	case SkillID:
		return s.finalLevel(s.bestFast(entity, requirePoints, excludes))
	default:
		level := entity.ResolveAttributeCurrent(s.Type())
		if ruleOf20 {
			level = level.Min(fxp.Twenty)
		}
		if entity.SheetSettings.UseHalfStatDefaults {
			level = level.Div(fxp.Two).Trunc() + fxp.Five
		}
		return s.finalLevel(level)
	}
}

func (s *SkillDefault) bestFast(entity *Entity, requirePoints bool, excludes map[string]bool) fxp.Int {
	best := fxp.Min
	for _, sk := range entity.SkillNamed(s.Name, s.Specialization, requirePoints, excludes) {
		if best < sk.LevelData.Level {
			best = sk.LevelData.Level
		}
	}
	return best
}

func (s *SkillDefault) finalLevel(level fxp.Int) fxp.Int {
	if level != fxp.Min {
		level += s.Modifier
	}
	return level
}
